programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -898,83 +880,948 @@ fn cpi_common<S: SyscallInvokeSigned>(
+            let data_addr = accounts_addr + accounts_len;
+
+            let ins = Instruction {
+                program_id: self.program_id,
+                accounts: unsafe {
+                    Vec::from_raw_parts(
+                        accounts_addr as *mut _,
+                        self.accounts.len(),
+                        self.accounts.len(),
+                    )
+                },
+                data: unsafe {
+                    Vec::from_raw_parts(data_addr as *mut _, self.data.len(), self.data.len())
+                },
+            };
+
+            unsafe {
+                ptr::write_unaligned(data.as_mut_ptr().cast(), ins);
+                data[accounts_addr - vm_addr..][..accounts_len].copy_from_slice(
+                    slice::from_raw_parts(self.accounts.as_ptr().cast(), accounts_len),
                 );
+                data[data_addr - vm_addr..].copy_from_slice(&self.data);
             }
-            to_slice.copy_from_slice(from_slice);
+
+            let region = MemoryRegion::new_writable(data.as_mut_slice(), vm_addr as u64);
+            (data, region)
         }
     }
 
-    Ok(SUCCESS)
+    fn mock_signers(signers: &[&[u8]], vm_addr: u64) -> (Vec<u8>, MemoryRegion) {
+        let slice_size = mem::size_of::<&[()]>();
+        let size = signers
+            .iter()
+            .fold(slice_size, |size, signer| size + slice_size + signer.len());
+
+        let vm_addr = vm_addr as usize;
+        let mut slices_addr = vm_addr + slice_size;
+
+        let mut data = vec![0; size];
+        unsafe {
+            ptr::write_unaligned(
+                data.as_mut_ptr().cast(),
+                slice::from_raw_parts::<&[&[u8]]>(slices_addr as *const _, signers.len()),
+            );
+        }
+
+        let mut signers_addr = slices_addr + signers.len() * slice_size;
+
+        for signer in signers {
+            unsafe {
+                ptr::write_unaligned(
+                    (data.as_mut_ptr() as usize + slices_addr - vm_addr) as *mut _,
+                    slice::from_raw_parts::<&[u8]>(signers_addr as *const _, signer.len()),
+                );
+            }
+            slices_addr += slice_size;
+            signers_addr += signer.len();
+        }
+
+        let slices_addr = vm_addr + slice_size;
+        let mut signers_addr = slices_addr + signers.len() * slice_size;
+        for signer in signers {
+            data[signers_addr - vm_addr..][..signer.len()].copy_from_slice(signer);
+            signers_addr += signer.len();
+        }
+
+        let region = MemoryRegion::new_writable(data.as_mut_slice(), vm_addr as u64);
+        (data, region)
+    }
+
+    struct MockAccountInfo<'a> {
+        pub key: Pubkey,
+        pub is_signer: bool,
+        pub is_writable: bool,
+        lamports: u64,
+        data: &'a [u8],
+        owner: Pubkey,
+        executable: bool,
+        rent_epoch: Epoch,
+    }
+
+    impl<'a> MockAccountInfo<'a> {
```

Comment:
how about a constructor or `from_account()` method?

Reply:
called it new(key, account) since it also needs the key